api/dataset_api.go

```diff
@@ -53,94 +54,108 @@ type JobLinks struct {
 }
 
 // GetInstance asks the Dataset API for the details for instanceID
-func (api *DatasetAPI) GetInstance(instanceID string) (Instance, error) {
+func (api *DatasetAPI) GetInstance(ctx context.Context, instanceID string) (instance Instance, err error, isFatal bool) {
 	path := api.url + "/instances/" + instanceID
 	logData := log.Data{"path": path, "instanceID": instanceID}
-	jsonBody, httpCode, err := api.get(path, nil)
+	jsonBody, httpCode, err := api.get(ctx, path, nil)
 	logData["httpCode"] = httpCode
-	if httpCode == http.StatusNotFound {
-		return Instance{}, nil
-	}
 	if err == nil && httpCode != http.StatusOK {
 		err = errors.New("Bad response while getting dataset instance")
+		if httpCode < http.StatusInternalServerError {
+			isFatal = true
+		}
+	} else if err != nil {
+		isFatal = true
 	}
 	if err != nil {
 		log.ErrorC("GetInstance get", err, logData)
-		return Instance{}, err
+		return
 	}
 	logData["jsonBody"] = jsonBody
 
-	var instance Instance
-	if err := json.Unmarshal(jsonBody, &instance); err != nil {
+	if err = json.Unmarshal(jsonBody, &instance); err != nil {
 		log.ErrorC("GetInstance unmarshall", err, logData)
-		return Instance{}, err
+		isFatal = true
 	}
-	return instance, nil
+	return
 }
 
 // GetInstances asks the Dataset API for all instances filtered by vars
-func (api *DatasetAPI) GetInstances(vars url.Values) ([]Instance, error) {
+func (api *DatasetAPI) GetInstances(ctx context.Context, vars url.Values) (instances []Instance, err error, isFatal bool) {
 	path := api.url + "/instances"
 	logData := log.Data{"path": path}
-	jsonBody, httpCode, err := api.get(path, vars)
+	jsonBody, httpCode, err := api.get(ctx, path, vars)
 	logData["httpCode"] = httpCode
 	if err == nil && httpCode != http.StatusOK {
 		err = errors.New("Bad response while getting dataset instances")
```

Comment:
Probably beyond the scope of this change but I think we need a team discussion on how we handle errors - for example using "github.com/pkg/errors" and making use of `errors.Wrap(...)` and passing the error back up the chain until something deals with it.